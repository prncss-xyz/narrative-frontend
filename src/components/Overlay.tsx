import { useClickOutside } from "../hooks/clickOutside";
import { Box, Flex } from "./basics";

export function Overlay({
  visible,
  setVisible,
  children,
  ...props
}: {
  visible: boolean;
  setVisible: () => void;
  children: React.ReactNode;
  [prop: string]: unknown; // TODO: could be more restrictive
}) {
  const ref = useClickOutside(setVisible);
  if (!visible) return null;
  return (
    <>
      <Box
        backgroundColor="black"
        css={{
          position: "fixed",
          left: "0px",
          top: "0px",
          zIndex: 10,
          opacity: "50%",
        }}
        height="100vh"
        width="100vw"
      />
      <Flex
        flexDirection="column"
        css={{
          position: "fixed",
          left: "0px",
          top: "0px",
          zIndex: 10,
        }}
        height="100vh"
        width="100vw"
      >
        <Flex justifyContent="center" height="100%" width="100%">
          <Flex
            justifyContent="center"
            alignItems="center"
            color="text"
            flexDirection="column"
          >
            <Box
              ref={ref}
              backgroundColor="tone1"
              p={2}
              borderStyle="solid"
              {...props}
            >
              {children}
            </Box>
          </Flex>
        </Flex>
      </Flex>
    </>
  );
}
