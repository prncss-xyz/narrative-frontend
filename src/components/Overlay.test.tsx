import { fireEvent, render, screen } from "@testing-library/react";
import { Overlay } from "./Overlay";

describe("Overlay", () => {
  describe("closed", () => {
    it("should not display child component", () => {
      const { container } = render(
        <div data-testid="closed-outside">
          <Overlay
            visible={false}
            setVisible={() => {
              /* do nothing */
            }}
          >
            <div>closed-inside</div>
          </Overlay>
        </div>
      );
      expect(screen.queryByText("closed-inside")).toBeNull();
      expect(container).toMatchSnapshot();
    });
  });
  describe("opened", () => {
    it("should call handler when clicked outside", () => {
      const spy = vi.fn();
      const { container } = render(
        <div data-testid="opened-outside">
          <Overlay visible={true} setVisible={spy}>
            <div>opened-inside</div>
          </Overlay>
        </div>
      );
      screen.getByText("opened-inside");
      fireEvent.click(screen.getByTestId("opened-outside"));
      expect(spy).toBeCalledTimes(1);
      expect(container).toMatchSnapshot();
    });
  });
});
