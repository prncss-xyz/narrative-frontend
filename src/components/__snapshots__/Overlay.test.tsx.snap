// Vitest Snapshot v1

exports[`Overlay > closed > should not display child component 1`] = `
<div>
  <div
    data-testid="closed-outside"
  />
</div>
`;

exports[`Overlay > opened > should call handler when clicked outside 1`] = `
<div>
  <div
    data-testid="opened-outside"
  >
    <div
      class="css-5uqfag-Box-Overlay e2i8k6n3"
      height="100vh"
      width="100vw"
    />
    <div
      class="css-1q8bug2-Box-Flex-Overlay e2i8k6n2"
      height="100vh"
      width="100vw"
    >
      <div
        class="css-18ya61d-Box-Flex e2i8k6n2"
        height="100%"
        width="100%"
      >
        <div
          class="css-15qjs6w-Box-Flex e2i8k6n2"
          color="text"
        >
          <div
            class="css-cex8gu-Box e2i8k6n3"
          >
            <div>
              opened-inside
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
